compiler/pipes/extract-resumable-calls.cpp

```diff
@@ -5,78 +5,160 @@
   } else if (auto if_cond = vertex.try_as<op_if>()) {
-    resumable_func_call = &if_cond->cond();
+    resumable_expr = &if_cond->cond();
   }
 
-  if (!resumable_func_call || !*resumable_func_call) {
+  if (!resumable_expr || !*resumable_expr) {
     return {};
   }
 
-  skip_conv_and_sets(resumable_func_call);
-
-  if ((*resumable_func_call)->type() != op_func_call || !(*resumable_func_call).as<op_func_call>()->func_id->is_resumable) {
+  resumable_expr = skip_conv_and_sets(resumable_expr);
+  if (!is_resumable_expr(*resumable_expr)) {
     return {};
   }
-
-  return resumable_func_call;
+  return resumable_expr;
 }
 
-VertexAdaptor<op_var> ExtractResumableCallsPass::make_temp_resumable_var(const TypeData *type) {
-  auto temp_var = VertexAdaptor<op_var>::create();
+std::pair<VertexAdaptor<op_move>, VertexAdaptor<op_set>> ExtractResumableCallsPass::make_temp_resumable_var(VertexPtr init) noexcept {
+  auto temp_var = VertexAdaptor<op_var>::create().set_location(init);
   temp_var->str_val = gen_unique_name("resumable_temp_var");
   temp_var->var_id = G->create_local_var(stage::get_function(), temp_var->str_val, VarData::var_local_t);
-  temp_var->var_id->tinf_node.copy_type_from(type);
+  temp_var->var_id->tinf_node.copy_type_from(tinf::get_type(init));
+  auto set_op = VertexAdaptor<op_set>::create(temp_var.clone().set_rl_type(val_l), init).set_rl_type(val_none).set_location(init);
+  auto move_op = VertexAdaptor<op_move>::create(temp_var.set_rl_type(val_r)).set_rl_type(val_r).set_location(init);
+  return {move_op, set_op};
+}
+
+VertexPtr ExtractResumableCallsPass::replace_set_ternary(VertexAdaptor<op_set_modify> set_vertex, VertexAdaptor<op_ternary> rhs_ternary) noexcept {
+  auto set_true_case = set_vertex.clone();
+  *skip_conv_and_sets(&set_true_case->rhs()) = rhs_ternary->true_expr();
+
+  auto set_false_case = set_vertex.clone();
+  *skip_conv_and_sets(&set_false_case->rhs()) = rhs_ternary->false_expr();
+
+  auto ternary_replacer = VertexAdaptor<op_if>::create(rhs_ternary->cond(), GenTree::embrace(set_true_case), GenTree::embrace(set_false_case));
+  return GenTree::embrace(ternary_replacer.set_rl_type(val_none).set_location(set_vertex));
+}
+
+VertexPtr ExtractResumableCallsPass::replace_set_logical_operation(VertexAdaptor<op_set_modify> set_vertex, VertexAdaptor<meta_op_binary> operation) noexcept {
+  auto temp_var = make_temp_resumable_var(operation->lhs());
+
+  auto set_if_lhs_true = set_vertex.clone();
+  auto set_if_lhs_false = set_vertex.clone();
+
+  switch (operation->type()) {
+    case op_log_or:
+      *skip_conv_and_sets(&set_if_lhs_true->rhs()) = temp_var.first;
+      *skip_conv_and_sets(&set_if_lhs_false->rhs()) = operation->rhs();
+      break;
+    case op_log_and:
+      *skip_conv_and_sets(&set_if_lhs_false->rhs()) = temp_var.first;
+      *skip_conv_and_sets(&set_if_lhs_true->rhs()) = operation->rhs();
+      break;
+    default:
+      kphp_assert(!"unexpected type");
```

Comment:
We usually write kphp_assert(0 && "message")?

Reply:
Interesting :)